File: 1236-n-th-tribonacci-number/1236-n-th-tribonacci-number.java
class Solution {
    public int tribonacci(int n) {
        Map<Integer, Integer> map = new HashMap<>();
        return recursion(n, map);
        
    }

    int recursion(int n, Map<Integer, Integer> memo) {

        if (memo.containsKey(n)) {
            return memo.get(n);
        }
        int result;
        if (n == 0) {
            result =  0;
        } else if (n == 1 || n == 2) {
            result = 1;
        } else {
            result = recursion(n-1, memo) + recursion(n-2, memo) + recursion(n-3, memo);
        }

        memo.put(n, result);
        return result;
    }
}

File: 1236-n-th-tribonacci-number/1236-n-th-tribonacci-number.py
class Solution:
    def tribonacci(self, n: int) -> int:
        if n == 0:
            return 0
        if n == 1 or n == 2:
            return 1

        memo = [0] * (n+1)
        memo[1] = 1
        memo[2] = 1
        
        for i in range(3, n+1):
            memo[i] = memo[i-1] + memo[i-2] + memo[i-3]
        return memo[n]
                